File: leetcode/322.coin-change.py
# @leet start
from math import inf


class Solution:
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp = [inf] * (amount + 1)
        dp[0] = 0

        for coin in coins:
            for i in range(coin, amount + 1):
                dp[i] = min(dp[i], dp[i - coin] + 1)

        return dp[amount] if dp[amount] < inf else -1


# @leet end

File: leetcode/322.coin-change.ts
// @leet start
function coinChange(coins: number[], amount: number): number {
  const dp = new Array(amount + 1).fill(Infinity);
  dp[0] = 0;

  for (const coin of coins) {
    for (let subAmount = coin; subAmount <= amount; subAmount++) {
      dp[subAmount] = Math.min(dp[subAmount], 1 + dp[subAmount - coin]);
    }
  }

  return dp[amount] < Infinity ? dp[amount] : -1;
}
// @leet end
